### PR TITLE
Issue #205

### PR DESCRIPTION
**Pull request by charlie0605**
Avatars are able to change name
image
image

### CODE REVIEW

**Comment by vickis** [comment]
wahoo!

**Comment by helenvarley** [comment]
how???

**Comment by charlie0605** [comment]
@helenvarley when the name is entered and you click on the edit button the name will edit

**Comment by helenvarley** [comment]
when the name is entered where? i don't see any field to enter a name - i'm looking in the avatar palette on https://limelight-qa.upstage.org.nz/hentest. maybe it's not been updated yet?

**Comment by charlie0605** [comment]
@helenvarley yes that might be the case

**Comment by helenvarley** [comment]
ok, then i'll wait for the order to test :)

File: app/assets/javascripts/channels/avatars.coffee [hunk 1]
```diff
@@ -64,6 +71,13 @@ hold = (data) ->
...
   btn.setAttribute 'disabled', 'disabled'
   btn.setAttribute 'title', "#{btn.getAttribute 'title'} (#{data.username})"
+  if avatarName[data.avatar_id] == undefined
+    avatarName[data.avatar_id] = data['name']
+  nameInput = document.querySelector '#editAvatarName'
+  nameInput.value = avatarName[data.avatar_id]
+  nameInput.removeAttribute 'disabled'
+  editBtn = document.querySelector '#editNameBtn'
+  editBtn.removeAttribute 'disabled'
   if `data.avatar_id == window.holdWait`
```
Comment (thongteav) [inline app/assets/javascripts/channels/avatars.coffee]: What would happen to another connected user when you drop an avatar? Can they still change the name of the avatar they're using?

File: app/views/theatre/performance.html.erb [hunk 2]
```diff
@@ -69,6 +69,10 @@
             </div>
           </div>
           <div class="avatar-functions row">
+            <div class="edit-name" style="float:left; margin-left: 20px;">
+              <input placeholder="name" disabled id="editAvatarName">
+                <button class="btn btn-default" id="editNameBtn" disabled="disabled">Edit</button>
+            </div>
```
Comment (thongteav) [inline app/views/theatre/performance.html.erb]: Not too sure if you should do styling here. Maybe do it in the stylesheet?

File: app/assets/javascripts/channels/avatars.coffee (hunk 1 above)
Comment (helenvarley) [reply app/assets/javascripts/channels/avatars.coffee]: they should be able to. it should be that whenever you are holding an avatar, you can change its name; independent of anything else that might be happening at the same time on the same stage.

**Comment by rohrlach** [comment]
@charlie0605 what's the status on this? Are you ready for me to review?

**Comment by charlie0605** [comment]
@rohrlach im still working on it, when the edit button is pressed the name does not change on the canvas unless its followed by a click on the canvas to move the avatar.(This is when 2 user is present).

**Comment by charlie0605** [comment]
@rohrlach can you give me some feedback on if this approach is the right approach of doing this issue?

**Comment by rohrlach** [comment]
@charlie0605 I don't see anything immediately wrong with the code.  Can you confirm that it works? IE: If you change the name on one machine, the name is changed on another machine?

**Comment by rohrlach** [comment]
One major problem I can see is that other users do not see the the name change until after the avatar is moved.  This suggests that either:

- A message is not sent to other users to say "Change the name"
or
- There is a message being sent but it does not work

**Comment by charlie0605** [comment]
@rohrlach there is that problem that you mention and also there is a problem when there are multiple users editing the name at the same time.

**Comment by rohrlach** [comment]
@charlie0605 that should never happen. The idea is that only one player can hold a specific avatar at a time.  

I'll raise a new issue to sure up one-holder-at-a-time. But please amend this so that name changes are "instant"

**Comment by helenvarley** [comment]
agree with paul's comment above - a player should only be able to change the name of the avatar they are holding. as only one player can hold a specific avatar at a time, no-one else should be able to edit the name at the same time.

**Comment by rohrlach** [comment]
@charlie0605 let's try and get this done this week.  I'll endeavor to respond promptly so that once this is in, we can tag a release and be done for the year.

**Comment by rohrlach** [comment]
@charlie0605 there are some conflicts to resolve after the merging of @thongteav s PR. Work together on them.

**Comment by vickis** [comment]
yay yes agree - player can own avatar but not others (as above only when
held and only one player at a time can hold to avoid this and other
conflicts ;)

be great to get a release out for 2017 - what a christmas present

yay virtual eggnogs all round :D

**Comment by charlie0605** [comment]
@rohrlach I have committed my new code, and have tested it to make sure that it works with 2 users. When one user changes an avatar name it doesn't affect other users. Also, the name will change immediately to the other user.

File: app/assets/javascripts/channels/avatars.coffee [hunk 3]
```diff
@@ -125,8 +125,7 @@ nameToggle = (data) ->
 
 editName = (data) ->
   avatarName = data.names
-  nameInput = document.querySelector '#editAvatarName'
-  avatarName[data.avatar_id] = nameInput.value
+  avatarName[data.avatar_id] = data.nickname
```
Comment (rohrlach) [inline app/assets/javascripts/channels/avatars.coffee]: That makes more sense 😄

**Comment by thongteav** [comment]
Testing two accounts, one on firefox, another on chrome.
Tests covered:
- A player can change the name of the avatar
changing name
- Name in the textfield changes when a new avatar is selected
name changes in textfield when avatar is changed
- Name changes instantly when edit button is pressed
name changes instantly
- Changes in the name reflected on other connected clients
name changes for other connected client

**Pull request merged**